[Console] Replace global `GET /_mapping` request with `GET <index>/_mapping`

darnautov opened the pull request:
## Summary

### Notes for reviewers 

- Currently, autocomplete suggestions for fields don't work with wildcards and data streams due to the [bug](https://github.com/elastic/kibana/issues/149496) in the `main`. It should be addressed separately. 

### How to test

In order to spot the loading behaviour, ideally you should create an index with a heavy mappings definition.  
Afterwards, write a query that requires a field from this index, e.g.:

```
GET <my-index>/_search
{
  "aggs": {
    "my_agg": {
      "terms": {
        "field": "",
        "size": 10
      }
    }
  }
}
```

Place a cursor next to the `field` property, it should trigger mappings fetch. After that, the mappings definition for this index will be cached and accessed synchronously. 

You can also open the browser's dev tools and enable Network throttling. It allows noticing loading behaviour for any index. 

--------------------

Resolves https://github.com/elastic/kibana/issues/146855

Instead of fetching all mappings upfront, requests mapping definition on demand per index according to the cursor position. 

Considering there is a maximum response size limit of 10MB in the `/autocomplete_entities` endpoint, field autocompletion wasn't working at all if the overall mappings definition exceeded this size.  Retrieving mappings per index tackles this and improves the init time. 

![Jan-25-2023 17-16-31](https://user-images.githubusercontent.com/5236598/214616790-4954d005-e56f-49f9-be6d-435c076270a8.gif)

### Checklist

- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

jloleysens commented on src/plugins/console/public/lib/autocomplete/autocomplete.ts:
```diff
@@ -1123,80 +1126,105 @@ export default function ({
+
+        if (context.asyncResultsState) {
+          // TODO indicate mappings fetching is in progress
+          context.asyncResultsState.results$.subscribe((r) => {
```
Is it possible that we have a memory leak here: calling `.subscribe` on new observables without unsubscribing?

darnautov replied:
replaced observable with a promise in https://github.com/elastic/kibana/pull/147770/commits/ea68a57143f218a5b54e61c066a31591125e7500

jloleysens commented on src/plugins/console/public/services/autocomplete.ts:
```diff
@@ -53,7 +53,11 @@ export class AutocompleteInfo {
       case ENTITIES.FIELDS:
-        return this.mapping.getMappings(context.indices, context.types);
+        return this.mapping.getMappings(
+          Object.getPrototypeOf(context),
```
Why are we using `Object.getPrototypeOf` here?

darnautov replied:
Because the parent (autocomplete) context we need to update is assigned to a prototype [here](https://github.com/elastic/kibana/blob/6bafb59fd5cc6cac55f07bd42df97e2ece0633e9/src/plugins/console/public/lib/autocomplete/engine.js#L38)

jloleysens reviewed (commented):
Great work @darnautov ! I'm glad this improvement is progressing. I tested locally and I think I could get the intended behaviour but saw a couple weird things too:


---

Saw this error in the browser console, not sure whether this is new? Perhaps a result of the WIP environment.

<img width="873" alt="Screenshot 2022-12-19 at 17 09 58" src="https://user-images.githubusercontent.com/8155004/208469780-9e57a324-9cfc-4980-98f0-f9eb6be40f2c.png">

---

In my local testing, it looks like getMappings can get called _on every cursor move_. If the cursor is in a position where it needs to look up mappings. This can result in quite a degraded experience. I am guessing the idea will be to cache responses after the initial fetch. But there is a possibility that before results are cached a lot of requests for the same mappings are fired.

https://user-images.githubusercontent.com/8155004/208474480-9646433a-0f52-41b1-8210-ff25f987f30e.mov

---

I'm guessing the mappings fetch performance fix is being worked on in a separate PR?

darnautov commented on src/plugins/console/public/lib/autocomplete_entities/mapping.ts:
```diff
@@ -70,22 +73,133 @@ function getFieldNamesFromFieldMapping(
+
+        const subj$ = new Subject<ResultTerm[]>();
+
+        autoCompleteContext.asyncResultsState!.results$ = subj$.asObservable();
```
Valid point! Indeed, using observables in this case is an overkill, replaced with promise in https://github.com/elastic/kibana/pull/147770/commits/ea68a57143f218a5b54e61c066a31591125e7500

darnautov commented:
> Great work @darnautov ! I'm glad this improvement is progressing. I tested locally and I think I could get the intended behaviour but saw a couple weird things too:
> 
> Saw this error in the browser console, not sure whether this is new? Perhaps a result of the WIP environment.
> 
> <img alt="Screenshot 2022-12-19 at 17 09 58" width="873" src="https://user-images.githubusercontent.com/8155004/208469780-9e57a324-9cfc-4980-98f0-f9eb6be40f2c.png">
> 
> In my local testing, it looks like getMappings can get called _on every cursor move_. If the cursor is in a position where it needs to look up mappings. This can result in quite a degraded experience. I am guessing the idea will be to cache responses after the initial fetch. But there is a possibility that before results are cached a lot of requests for the same mappings are fired.
> 
>  Screen.Recording.2022-12-19.at.17.33.25.mov 
> I'm guessing the mappings fetch performance fix is being worked on in a separate PR?

thanks for testing @jloleysens, I fixed the issues you mentioned in the latest commits.

jloleysens reviewed (commented):
@darnautov thanks for addressing a lot of my feedback. I tested a bit more closely this time and have a few notes:

---

Data streams seem to not be extracting the correct underlying value, looks like due to the index we got from console not matching the backing index name. E.g.:

`kibana_sample_data_logs` => `.ds-kibana_sample_data_logs-2023.01.03-000001`.

For reference, here is the full mapping response for the "Sample web logs".

```json
{".ds-kibana_sample_data_logs-2023.01.03-000001":{"mappings":{"_data_stream_timestamp":{"enabled":true},"properties":{"@timestamp":{"type":"date"},"agent":{"type":"text","fields":{"keyword":{"type":"keyword","ignore_above":256}}},"bytes":{"type":"long"},"bytes_counter":{"type":"long","time_series_metric":"counter"},"bytes_gauge":{"type":"long","time_series_metric":"gauge"},"clientip":{"type":"ip"},"event":{"properties":{"dataset":{"type":"keyword"}}},"extension":{"type":"text","fields":{"keyword":{"type":"keyword","ignore_above":256}}},"geo":{"properties":{"coordinates":{"type":"geo_point"},"dest":{"type":"keyword"},"src":{"type":"keyword"},"srcdest":{"type":"keyword"}}},"host":{"type":"text","fields":{"keyword":{"type":"keyword","ignore_above":256}}},"index":{"type":"text","fields":{"keyword":{"type":"keyword","ignore_above":256}}},"ip":{"type":"ip"},"ip_range":{"type":"ip_range"},"machine":{"properties":{"os":{"type":"text","fields":{"keyword":{"type":"keyword","ignore_above":256}}},"ram":{"type":"long"}}},"memory":{"type":"double"},"message":{"type":"text","fields":{"keyword":{"type":"keyword","ignore_above":256}}},"phpmemory":{"type":"long"},"referer":{"type":"keyword"},"request":{"type":"text","fields":{"keyword":{"type":"keyword","time_series_dimension":true}}},"response":{"type":"text","fields":{"keyword":{"type":"keyword","ignore_above":256}}},"tags":{"type":"text","fields":{"keyword":{"type":"keyword","ignore_above":256}}},"timestamp":{"type":"alias","path":"@timestamp"},"timestamp_range":{"type":"date_range"},"url":{"type":"text","fields":{"keyword":{"type":"keyword","ignore_above":256}}},"utc_time":{"type":"date"}}}}}
```

---

Similar to data stream case above, we are not getting completions for wild cards `*`. Aliases seem to work correctly!

---

Do you think it is possible to extract the performance fix to a separate PR so that we can merge that ASAP?

---

Is it possible to make completion suggestions wait for the request to finish?

darnautov commented:
hey @jloleysens,

Just tested on `main`, both data streams and wild cards don't seem to work 🤷🏻 
Do you think it's worth fixing in this PR? 

> Do you think it is possible to extract the performance fix to a separate PR so that we can merge that ASAP?

It's been fixed in https://github.com/elastic/kibana/pull/148328 and backported 

> Is it possible to make completion suggestions wait for the request to finish?

Do you mean showing some loading indicator?

szabosteve reviewed (approved):
One small suggestion, otherwise UI text LGTM.

jloleysens reviewed (changes requested):
Great work @darnautov . I tested it locally and it works as expected; but I am seeing a few issues around caching. I think we should address the first one, the other two are not as important, but would be nice to address before merging:

---

I'm seeing the following behaviour:

1. Enter an index name, like `GET my-index-1/_search` and trigger autocomplete for fields (like doing agg terms)
2. Create a new index: `PUT my-index-2` with some mappings
3. Enter the new index name `GET my-index-2/_search` and trigger autocomplete for fields
4. Enter the first index name like `GET my-index-1/_search` and trigger autocomplete for fields. Now there is no autocomplete for fields -- there is only field autocomplete for the `my-index-2` fields.

From what I can tell, the `loadingState` and `perIndexTypes` pieces of state are getting out of sync because the `perIndexTypes` is overwritten to `{}` whenever `loadMappings` is called, but `loadingState` still has `my-index-1` as `true` for loading state so re-fetch is not triggered and the value was lost by calling `loadMappings` so no fields are found.

I think we should "merge" the new mappings with the existing `perIndexTypes` to allow the cache to grow with new objects. We could also consider using a `LRU` to expire values (I'd consider this as another PR though). Also it _might_ be simpler to only have `perIndexTypes` as the source of truth for when to fetch and clean up values in `loadingState` as the request completes. WDYT?

---

👉🏻 if you enter an index name that does not exist:

```
GET index-name-with-typo/_search
{
  "aggs": {
    "test": {
      "terms": {
        "field": "",
        "size": 10
      }
    }
  }
}
```

...and trigger fetching of mappings by, for example, doing a terms agg. The empty mappings response is cached. This can create a stale cache issue if the `index-name-with-typo` is was created. Sequence is:

1. Trigger mappings fetch for non-existent index
5. Create the index (that didn't exist in step 1)
6. Trigger mappings fetch and see empty cached result

I don't expect this sequence to be very common, but it is possible and might be confusing for users.

I think we should have an error response (maybe 404?) from the server and not cache errored results. WDYT?

---

👉🏻 using wildcards `*` in index names can cause unexpected results, similar to above

My proposal, for now, is that we just do not cache fields fetched containing `*`. WDYT? Just always fetch...

Again, probably not the biggest issue, but would be nice to avoid unexpected UXs.

darnautov commented:
thanks for testing @jloleysens, addressed the issue with a cache overwrite in https://github.com/elastic/kibana/pull/147770/commits/aa2a8a576ea57d252eda1127e45920182233c2b5 and the one about stale cache in https://github.com/elastic/kibana/pull/147770/commits/bd821152adb619e804aea65bb97f4560029b57f8

jloleysens reviewed (approved):
@darnautov thanks for fixing the issue I found. I tested locally and found an infinite recursion was occurring when fetching fields for data streams (like the ones used for sample data).

I pushed a commit to fix this https://github.com/elastic/kibana/pull/147770/commits/e66a34a61587c52e0cb06307d5e301778733da3b, but I think it would be really nice to support them (I think you already have an issue to track this improvement). Let me know what you think of the change. Otherwise great work!

jloleysens commented:
Also, adding some tests for this functionality would be really nice!

alisonelizabeth reviewed (changes requested):
Thanks for working on this @darnautov! Awesome work 🎉 

I didn't look too closely at the code yet, but I do have a few general questions/comments.

 - As Jean-Louis commented, I think we really need to add tests for this functionality before merging.
- Have you done any performance tests, comparing the previous behavior to your change? I think it'd be helpful/interesting to include as part of the PR description.
- Can you update the PR description with some information on how to manually test this PR? Specifically, the best way to test the "loading" behavior.
- I'm not seeing autocomplete for some of the indices I added. Can you take a look?

On your branch:
<img width="409" alt="Screen Shot 2023-01-27 at 7 32 21 PM" src="https://user-images.githubusercontent.com/5226211/215507103-0904f62e-4106-4bfd-a27d-64a9d0909029.png">

(Confirmed I do have the indices)
<img width="542" alt="Screen Shot 2023-01-30 at 9 41 17 AM" src="https://user-images.githubusercontent.com/5226211/215507914-28ea396a-6ec2-424c-96fd-31b123ec54cf.png">

Expected behavior (on prod deployment):
<img width="424" alt="Screen Shot 2023-01-27 at 7 32 43 PM" src="https://user-images.githubusercontent.com/5226211/215507282-23cff3ba-3e37-47dd-a2f5-a3ec21510c90.png">

alisonelizabeth commented:
> Considering there is a maximum response size limit of 10MB in the /autocomplete_entities endpoint, field autocompletion wasn't working at all if the overall mappings definition exceeded this size. Retrieving mappings per index tackles this and improves the init time.

Do we still need the response size limit in place, with your change?


> I pushed a commit to fix this [e66a34a](https://github.com/elastic/kibana/commit/e66a34a61587c52e0cb06307d5e301778733da3b), but I think it would be really nice to support them (I think you already have an issue to track this improvement).

Can you share any follow-up issues you created in the PR description? Thanks!

jloleysens commented on src/plugins/console/public/application/models/legacy_core_editor/legacy_core_editor.ts:
```diff
@@ -412,7 +412,30 @@ export class LegacyCoreEditor implements CoreEditor {
+            let customAnnotation: Annotation;
+            return {
+              setAnnotation(text: string) {
+                const annotations = DO_NOT_USE_2.getAnnotations();
```
Haha yeah, this is part of a pattern I started to keep Ace-specific concepts outside of the autocomplete system so that we could migrate to a newer editor, like Monaco. We could probably rename the variable to describe what it actual is: `aceEditorSession` or something. And a comment that we should not be passing this to the autocompleter

darnautov commented:
thanks for the review @alisonelizabeth! 

> As Jean-Louis commented, I think we really need to add tests for this functionality before merging.

I've updated existing tests and added more test cases to cover new logic 

> Have you done any performance tests, comparing the previous behavior to your change? I think it'd be helpful/interesting to include as part of the PR description.

I haven't created any reports regarding the performance, but basically, the Console app doesn't load the mappings definition on start, which increate start-up time, especially for big clusters where the mappings definition exceeds 10 MB. It's going to vary based on the mapping size, but in any case, it's more optimal to not load everything upfront. 

> Can you update the PR description with some information on how to manually test this PR? Specifically, the best way to test the "loading" behavior.

Done ✅ Please let me know if any details are missing. 

> I'm not seeing autocomplete for some of the indices I added. Can you take a look?

Thanks for spotting this! Fixed in https://github.com/elastic/kibana/pull/147770/commits/094b5a82a3fa1a1612599e12eac1bd05ff5d44b2. I just reverted the change I had introduced at the beginning of this work, assuming it's redundant 😬 

> Do we still need the response size limit in place, with your change?

Yes, it's still needed. The user can potentially write a wildcard that matches a vast number of indices. 

> Can you share any follow-up issues you created in the PR description? Thanks!

Updated the PR description!

darnautov commented:
@elasticmachine merge upstream

darnautov commented:
@elasticmachine merge upstream

yuliacech reviewed (commented):
Hi @darnautov, I've tested locally and was able to still see several repeating requests going out for the same autocomplete requests. Also it seems like no mappings were returned for my test index. I'm attaching a screen recording below. 
I'll also have a look at a Cloud deployment to test there as well. 

https://user-images.githubusercontent.com/6585477/217000579-e18456f7-5d8b-452b-8911-cb0029cf04ea.mov

darnautov commented:
hi @yuliacech, in the video I spotted you have a typo in an index name. So I presume the index doesn't exist and the autocompletion code tries to retrieve the mapping. It's been done deliberately because the user can create an index in between.

darnautov commented:
@yuliacech but there was also an issue with the way index caching clashed with the mappings checked. thanks for spotting this! fixed in https://github.com/elastic/kibana/pull/147770/commits/4a7d4566fe857a006d6d9486699e8b6453d3d873

yuliacech reviewed (approved):
Thanks a lot for this fix, @darnautov! 
Code changes LGTM,  appreciate the addition of tests 👍 tested locally and in the cloud deployment, works as expected. No more repeatable requests.

darnautov commented:
@elasticmachine merge upstream

alisonelizabeth reviewed (approved):
Thanks for addressing my feedback! @yuliacech reviewed on my behalf, but approving to unblock merging.
